### PR TITLE
llvm 13.0.0

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

### CODE REVIEW

**Comment by carlocab** [comment]
GCC 5 patch will need updating. CC @danielnachun

**Comment by Bo98** [comment]
The patch shouldn't be necessary in LLVM 13 since it's supposed to have been fixed upstream.

**Comment by carlocab** [comment]
> The patch shouldn't be necessary in LLVM 13 since it's supposed to have been fixed upstream.

LLDB still fails to build, so it seems this is not yet fixed. The fix referenced at the start of the bug report isn't complete.

**Comment by Bo98** [comment]
> The fix referenced at the start of the bug report isn't complete.

Ok I didn't notice the additional part. The addditional patch to `NativeRegisterContextLinux_x86_64::GetSyscallData` might still be needed but the rest of our patch, which was backporting an existing commit, should no longer be needed.

But the error doesn't appear to be complaining about that but rather `ThreadPlanStack` instead.

**Comment by carlocab** [comment]
Some of these look like they might need something other than a revision bump, but let's see.

Julia will fail to build with Xcode 13. JuliaLang/julia@4dea1c44bd428cd5309abd910a82f63d262833a6 needs backporting.

**Comment by carlocab** [comment]
`spirv-llvm-translator` has a hard requirement on LLVM 12 (i.e. it won't accept anything newer or older). Some of the others I've switched may not be absolutely necessary, but some are.

**Comment by carlocab** [comment]
```
In file included from /tmp/llvm-20210929-83167-i3pu42/llvm-project-13.0.0rc4.src/compiler-rt/lib/builtins/absvdi2.c:13:
In file included from /tmp/llvm-20210929-83167-i3pu42/llvm-project-13.0.0rc4.src/compiler-rt/lib/builtins/int_lib.h:93:
In file included from /tmp/llvm-20210929-83167-i3pu42/llvm-project-13.0.0rc4.src/llvm/build/lib/clang/13.0.0/include/limits.h:21:
In file included from /usr/include/limits.h:25:
/usr/include/features.h:367:12: fatal error: 'sys/cdefs.h' file not found
#  include <sys/cdefs.h>
           ^~~~~~~~~~~~~
1 error generated.
CMakeFiles/clang_rt.builtins-i386.dir/build.make:75: recipe for target 'CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o' failed
make[5]: *** [CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o] Error 1
make[5]: Leaving directory '/tmp/llvm-20210929-83167-i3pu42/llvm-project-13.0.0rc4.src/llvm/build/runtimes/builtins-bins'
CMakeFiles/Makefile2:439: recipe for target 'CMakeFiles/clang_rt.builtins-i386.dir/all' failed
make[4]: *** [CMakeFiles/clang_rt.builtins-i386.dir/all] Error 2
make[4]: Leaving directory '/tmp/llvm-20210929-83167-i3pu42/llvm-project-13.0.0rc4.src/llvm/build/runtimes/builtins-bins'
Makefile:135: recipe for target 'all' failed
make[3]: *** [all] Error 2
```
Ideas, anyone?

**Comment by danielnachun** [comment]
I think `sys/cdefs.h` is a part of `glibc`, so I do find it odd that it can't find those headers.  Nonetheless I think adding that include path is probably fine.  LLVM is relocatable so almost nobody will be building it from source.

File: Formula/llvm.rb [hunk 1]
```diff
@@ -158,6 +142,8 @@ def install
     end
 
     if OS.linux?
+      # The runtimes build can't find Glibc headers for some reason
+      ENV.append "CPPFLAGS", "-isystem /usr/include/x86_64-linux-gnu"
```
Comment (carlocab) [inline Formula/llvm.rb:146]: Still getting missing headers on Linux even after adding this.

**Comment by carlocab** [comment]
LLVM 13 is out. Still haven't worked out how to build it on Linux. Going to try Homebrew GCC.

**Comment by danielnachun** [comment]
It looks like brewed `gcc` doesn't solve our failure on LLVM 13.  I've done some testing and `ENV.append "CFLAGS", "/usr/include/x86_64-linux-gnu"` does fix the error about missing `sys/cdefs.h`, but then it fails with another error about missing `gnu/stubs-32.h`.  I believe this happens because `stubs.h` checks if `__x86_64__` is defined, and if so it is supposed to look for `gnu/stubs-64.h`, which we do have.

@chandlerc has anything changed with how `clang` finds `libc` headers on Linux?  We're hitting a failure when trying to build `compiler-rt` with the freshly built `clang` fails because it doesn't know how to find the architecture-specific headers for `libc`, and if we add a `CFLAG` to help it find them, it apparently doesn't define the architecture correctly and looks for 32-bit headers on a 64-bit system.

**Comment by carlocab** [comment]
This is probably an issue with the just-built `clang` rather than in `compiler-rt`. One way to fix this might be to move `compiler-rt` to the `projects` array on Linux (this will break the build on macOS, unless we drop PGO). It'll tie us even more tightly to Homebrew GCC, however, since `compiler-rt` doesn't build on GCC 5 (at least, it didn't on LLVM 12).

**Comment by Bo98** [comment]
> This is probably an issue with the just-built `clang` rather than in `compiler-rt`. One way to fix this might be to move `compiler-rt` to the `projects` array on Linux (this will break the build on macOS, unless we drop PGO).

If this is indicating the just-built clang is broken, then we should be addressing that. Can we get the standard include path printed somewhere to have a look?

**Comment by carlocab** [comment]
> If this is indicating the just-built clang is broken, then we should be addressing that. Can we get the standard include path printed somewhere to have a look?

I'm not aware of a way to do this, but if we can get the build to finish we can poke at bottle to see if there's something wrong with `clang`.

**Comment by danielnachun** [comment]
We might be able to check if we build with `--debug` and look at the just-built `clang` in the debug shell. Also I am looking into the GCC 5 build failure for `compiler-rt`, because that may be something upstream can address, and this would allow us to have a PGO build on Linux.

**Comment by chandlerc** [comment]
Very happy to help here -- forgive my lack of knowledge of Homebrew -- how should I reproduce this?

It's not too hard to get `clang` to print out the paths used in header search, compiling with `-v` will do so (along with other debugging aids)...

Adding something like `/usr/include/x86_64-linux-gnu` to `-isystem` shouldn't be needed, Clang (and GCC) should automatically look there...

**Comment by danielnachun** [comment]
Thanks for offering to help @chandlerc.  You can run our CI environment in Docker with `docker run -it homebrew/ubuntu16.04:master` and then either check out this branch or edit the `llvm` formula to match what we are doing. 

Ideally I think we should still build with GCC 5 if LLVM 13 supports it.  Separately from the whether the projects still support GCC 5, both @carlocab and I found that the runtimes (`compiler-rt`, `libc++`, and `libc++abi` at least, not sure about `libunwind`) fail to build with GCC 5 even in LLVM 12.  In order to do a PGO build on Linux I think we'd have to get the runtimes to build with GCC 5 once before the final versions would be built with `clang`.  But I think the first priority here is to understand why `clang` can't find the `libc` headers.

**Comment by carlocab** [comment]
> Very happy to help here -- forgive my lack of knowledge of Homebrew -- how should I reproduce this?

I think you should be able to check out this PR branch and build from there. You might want to switch your Git remote at `$(brew --repo homebrew/core)` to https://github.com/Homebrew/homebrew-core first (and set `HOMEBREW_FORCE_HOMEBREW_ON_LINUX` to make sure `brew` doesn't try to move you back to Linuxbrew).

> It's not too hard to get `clang` to print out the paths used in header search, compiling with `-v` will do so (along with other debugging aids)...

I contemplated that, but it would generate an ungodly amount of logs. GitHub already refuses to show the compilation log from the previous build failure. I guess it's worth a try if we get stuck.

**Comment by carlocab** [comment]
Actually, why is this trying to build what presumably are i386 builtins?
```
CMakeFiles/clang_rt.builtins-i386.dir/build.make:75: recipe for target 'CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o' failed
make[5]: *** [CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o] Error 1
```
Did it always do that?

**Comment by danielnachun** [comment]
Perhaps that's needed if you want to cross-compile to i386?  Especially if you want to statically link the runtime, where you'd need a copy yourself to include in the final binary.

**Comment by danielnachun** [comment]
> We only need to build `compiler-rt` to get a PGO build going, so there's no need to work out `libc++[abi]`.

That's good to know.  I'm currently trying to build with `compiler-rt` as a project so I can post the exact error here for discussion.

**Comment by Bo98** [comment]
It's normal on macOS, but I don't think our Linux build ever built that. I'm guessing our GCC 5 might not support i386 but our brewed GCC does?

`COMPILER_RT_DEFAULT_TARGET_ONLY=ON` will make it only do x86_64 like before.

**Comment by danielnachun** [comment]
That error is the same whether we use brewed GCC or GCC 5, so I don't think that's the explanation.  I don't think we actually have a reason right now to use brewed GCC, as that is almost exclusively needed when we need a newer C++ compiler (or maybe Fortran).  The failure we're seeing now is in C code, and I wouldn't expect that to have changed much.

**Comment by danielnachun** [comment]
I just tried to do a build with `compiler-rt` as a project instead of a runtime: https://gist.github.com/71ad909db0954d2ef71231a0253e7447.  I think `compiler-rt` itself built and the build failed while building `libc++abi` as a runtime with the freshly built `clang`, because it could not find the freshly built `compiler-rt`.  I just noticed that this was set to build `libunwind` as project instead of a runtime, so I'm going to change `libunwind` back to a runtime as it originally was and see if that makes a difference.

**Comment by carlocab** [comment]
Looks like the macOS builds should be green.

I've added `COMPILER_RT_DEFAULT_TARGET_ONLY=ON` to `RUNTIMES_CMAKE_ARGS` and kicked off a build at https://github.com/carlocab/homebrew-personal/pull/106. Let's see if that helps.

**Comment by chandlerc** [comment]
The build worked fine for me on Debian testing...

I also ran the built clang to get its header search:
```
> clang -v -x c++ -o /dev/null /dev/null
...
#include <...> search starts here:
 /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10
 /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/x86_64-linux-gnu/c++/10
 /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/backward
 /home/linuxbrew/.linuxbrew/Cellar/llvm/13.0.0/lib/clang/13.0.0/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```

So it is searching `/usr/include/x86_64-linux-gnu`.

> I just tried to do a build with `compiler-rt` as a project

I'd hope to avoid this.

**Comment by chandlerc** [comment]
> I've added `COMPILER_RT_DEFAULT_TARGET_ONLY=ON` to `RUNTIMES_CMAKE_ARGS` and kicked off a build at [carlocab/homebrew-personal#106](https://github.com/carlocab/homebrew-personal/pull/106). Let's see if that helps.

This will likely help, but may not be the solution you want...

The compile command that fails is this:
```
/tmp/llvm-20211003-180167-ky2r1i/llvm-project-13.0.0.src/llvm/build/./bin/clang --target=x86_64-unknown-linux-gnu -DVISIBILITY_HIDDEN  -fpermissive -Wno-free-nonheap-object -O3 -DNDEBUG  -m32 -std=c11 -fPIC -f
no-builtin -fvisibility=hidden -MD -MT CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o -MF CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o.d -o CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o -c /tmp/llvm-20211003-180167-ky2r1i/llvm-project-13.0.0.src/compiler-rt/lib/builtins/absvdi2.c
```

The `-m32` there is key and causes it to need the 32-bit version of these headers...

Maybe can fix by installing some amount of the cross building packages on the system. For example, `crossbuild-essentials-i386`. This gives me a `/usr/i686-linux-gnu/include/` directory, which my `clang` searches (through a weird set of directories and `../`s) when I give a similar commandline flag.

This is all trying to build the runtimes that Clang wants to be able to cross compile. Clang is essentially always a cross compiler and so it tries to build these runtimes by default.

**Comment by carlocab** [comment]
> > I've added `COMPILER_RT_DEFAULT_TARGET_ONLY=ON` to `RUNTIMES_CMAKE_ARGS` and kicked off a build at [carlocab/homebrew-personal#106](https://github.com/carlocab/homebrew-personal/pull/106). Let's see if that helps.
> 
> This will likely help, but may not be the solution you want...

I don't think we've ever been able to build 32-bit targets in Linux CI, so at the very least this won't be a regression for any existing users.

I'm inclined to go with this for now (if it does work) so that we can ship the new version and then see if we can get cross-compiling to work in CI after.

**Comment by danielnachun** [comment]
> Maybe can fix by installing some amount of the cross building packages on the system. For example, `crossbuild-essentials-i386`. This gives me a `/usr/i686-linux-gnu/include/` directory, which my `clang` searches (through a weird set of directories and `../`s) when I give a similar commandline flag.
> 
> This is all trying to build the runtimes that Clang wants to be able to cross compile. Clang is essentially always a cross compiler and so it tries to build these runtimes by default.

I was guessing that installing the 32-bit `libc` headers would be a possible solution, though I don't understand why we didn't run into this problem before.  Maybe it failed silently in previous versions?  Regardless, this does pose a bit of challenge to us as we need to either make the 32-bit `libc` headers available or disable the ability to cross-compile to 32-bit, which may be limiting some potentially useful functionality.

One possible solution here is to package the `glibc` headers separately from the libraries.  In CI we want to use the host `glibc` for the shared libraries and executables, but the headers may be a different story.  I see in the `glibc` source archive that there is only a `stubs.h` and that it contains a comment which says that the stubs are generated by `make install`.  Arch Linux provides a 32-bit `glibc` for compiling as part of split package, and they actually build `glibc` twice, passing `--host=i686-pc-linux-gnu` the second time to get the 32-bit libraries and headers.  

I'm not sure what our general policy is in terms of support for building toolchains with cross-compile support.  If we can't test any cross-compiled binaries we wouldn't know if they are broken.  I'm not even sure if our GCC formulae can cross compile to `i386`.

**Comment by danielnachun** [comment]
Haha @carlocab I guess we posted at the same time so I didn't see your response about 32-bit targets.  I looked in the current LLVM 12 bottle and it only contains `x86_64` libraries for `libclang_rt`.  So I don't think our previous LLVM versions could have cross-compiled to 32-bit before either.

One other relevant point here is that architecture flags like `-m32` and `-m64` are normally stripped out by the `clang` wrapper in our superenv, but that superenv is bypassed when building runtimes because they are using the direct path to the newly built `clang` instead.  If `COMPILER_RT_DEFAULT_TARGET=ON` effectively achieves the same thing, then it would be consistent with how we normally build things.

**Comment by carlocab** [comment]
> One other relevant point here is that architecture flags like `-m32` and `-m64` are normally stripped out by the `clang` wrapper in our superenv, but that superenv is bypassed when building runtimes because they are using the direct path to the newly built `clang` instead. If `COMPILER_RT_DEFAULT_TARGET=ON` effectively achieves the same thing, then it would be consistent with how we normally build things.

Well, even if the runtimes build used the shims, we do set `ENV.permit_arch_flags` which allows the use of `-m32` and `-m64` flags.

**Comment by carlocab** [comment]
macOS CI is green. I'm going to try `COMPILER_RT_DEFAULT_TARGET_ONLY=ON` here.

**Comment by danielnachun** [comment]
From your personal tap run @carlocab:
```
 /tmp/llvmA13-20211004-11286-1kvudtr/llvm-project-13.0.0.src/llvm/build/./bin/clang 
 --target=x86_64-unknown-linux-gnu 
 -DVISIBILITY_HIDDEN  -fpermissive -Wno-free-nonheap-object -O3 -DNDEBUG  
 -m32 
 -std=c11 -fPIC -fno-builtin -fvisibility=hidden -MD 
 -MT CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o 
 -MF CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o.d 
 -o CMakeFiles/clang_rt.builtins-i386.dir/absvdi2.c.o 
 -c /tmp/llvmA13-20211004-11286-1kvudtr/llvm-project-13.0.0.src/compiler-rt/lib/builtins/absvdi2.c
```
It seems like setting the the target doesn't actually disable trying to build 32-bit run times, it just adds `--target=x86_64-unknown-linux-gnu` which only confuses things more.  I'm guessing that we need some other flag to disable the runtimes instead.  I'm still puzzled as to why in previous LLVM versions it seems the `i386` runtimes were ignored, and now they are not.

**Comment by Bo98** [comment]
> I've added `COMPILER_RT_DEFAULT_TARGET_ONLY=ON` to `RUNTIMES_CMAKE_ARGS` and kicked off a build at [carlocab/homebrew-personal#106](https://github.com/carlocab/homebrew-personal/pull/106). Let's see if that helps.

I think you want `BUILTINS_CMAKE_ARGS` for this instead.

**Comment by Bo98** [comment]
> I've added `COMPILER_RT_DEFAULT_TARGET_ONLY=ON` to `RUNTIMES_CMAKE_ARGS` and kicked off a build at [carlocab/homebrew-personal#106](https://github.com/carlocab/homebrew-personal/pull/106). Let's see if that helps.

This doesn't work because it needs to be passed to `BUILTINS_CMAKE_ARGS` instead.

Setting `LLVM_BUILTIN_TARGETS` also works (outside of `BUILTIN_CMAKE_ARGS`).

File: Formula/ldc.rb [hunk 2]
```diff
@@ -69,7 +74,7 @@ def install
 
       if OS.mac?
         # Workaround for https://github.com/ldc-developers/ldc/issues/3670
-        cp Formula["llvm"].opt_lib/"libLLVM.dylib", lib/"libLLVM.dylib"
+        cp llvm.opt_lib/"libLLVM.dylib", lib/"libLLVM.dylib"
```
Comment (carlocab) [inline Formula/ldc.rb:77]: I should replace this workaround with either a symlink, or setting `CMAKE_INSTALL_RPATH`. We already have a runtime dependency on LLVM anyway. Will consider doing that if I have to rebase this.

File: Formula/llvm.rb (hunk 1 above)
Comment (yurikoles) [reply Formula/llvm.rb]: >x86_64

Seems to be unportable. Also, on Linux there is an optional dependency on glibc [at line 43](https://github.com/Homebrew/homebrew-core/blob/a5c8d94db6a84362ab8471090eab58c13decf635/Formula/llvm.rb#L43) if it's installed:

```
depends_on "glibc" if Formula["glibc"].any_version_installed?
```

**Comment by carlocab** [comment]
Linux failed again with
```
tools/lldb/source/Target/CMakeFiles/lldbTarget.dir/build.make:78: recipe for target 'tools/lldb/source/Target/CMakeFiles/lldbTarget.dir/ABI.cpp.o' failed
make[2]: *** [tools/lldb/source/Target/CMakeFiles/lldbTarget.dir/ABI.cpp.o] Error 1
make[2]: Leaving directory '/tmp/llvm-20211005-88296-1cc4l96/llvm-project-13.0.0.src/llvm/build'
CMakeFiles/Makefile2:55863: recipe for target 'tools/lldb/source/Target/CMakeFiles/lldbTarget.dir/all' failed
make[1]: *** [tools/lldb/source/Target/CMakeFiles/lldbTarget.dir/all] Error 2
```
This worked on my testing elsewhere, but that was using GCC 10. I'm building with GCC on Linux.

**Comment by danielnachun** [comment]
@chandlerc the error @carlocab posted above is new as of LLVM 13, and prevents us from building with GCC 5.  Can you take a look?  It is probably related to this commit https://github.com/llvm/llvm-project/commit/6eb576dcff45fa53e8ceb30c7648b903c1413e05, though there were two others before it that were added after the LLVM 12 release.

**Comment by carlocab** [comment]
All builds are green. This should be ready to go very soon.

**Comment by carlocab** [comment]
Merging, since #80227 is waiting on this one.

**Comment by danielnachun** [comment]
I agree as it seems that LLVM (or at least LLDB) isn’t being tested with GCC 5 despite officially supporting it.  Given how demanding LLVM is in general of the host compiler I think using brewed GCC may be a better choice.  The biggest pain for Linux users with using having brewed GCC as a runtime for LLVM is that the GCC bottle is incorrectly made non-relocatable on Linux.  However this will be fixed by https://github.com/Homebrew/brew/pull/11899, which I will reopen and prioritize finishing.  From a maintainer perspective, the main issue is C++ runtime dependencies of LLVM needing to be built with brewed GCC.  I suspect a large chunk of that is through `mesa`, which requires brewed GCC regardless of LLVM.  If the set of remaining runtime dependents of LLVM which are not through `Mesa` is small, this won’t be too much work to fix them.

**Comment by danielnachun** [comment]
> Or, another alternative:
> 
> Build LLVM with a slightly older GCC (7?), make it a build-only dependency, and then add a `Requirement` that checks that the user has GCC runtime libraries that are new enough, which can presumably be satisfied by `brew install gcc` (or even `brew install gcc@7`). Would that work?

It's probably true that this would help to reduce the number of users who have to installed brewed `gcc` as a runtime dependency, but like I said above that will be less painful when `gcc` is a relocatable bottle which will be done soon, so it would just save some disk space. 

However I think the above solution still requires us to update all LLVM C++ dependents so they know to build with the newer GCC, as merely having some version of brewed GCC in the dependency tree does not mean it actually gets used to compile.  Note that this is necessary even if the formula is built with `clang`, because it still needs to prepend the newer `libstdc++` the `LDFLAGS` so that it doesn't try to use the system one which is too old.  Using some version of GCC between the host GCC in CI and newest version also creates more work in terms of keeping track of which version should be used.  @iMichka, @sjackman and I all had a pretty strong consenus that we wanted to limit the use of older versions of brewed GCC to the few cases where neither the host GCC nor the newest version of GCC would work, and that has been extremely rare.

These issues are making me wonder if in the long run on Linux it will make more sense for us to just start using brewed GCC for C++ on Linux.  On macOS using the host toolchain to build C++ binaries isn't too bad because Apple updates it for us regularly.  On Linux that's not the case, and it is the source of a lot of our difficulties with C++.  The fact that some users who happen to have a new enough host version of `libstdc++` would end up having to use an extra 225MB to install another copy of GCC to me is not worth all the trouble of trying to use the host GCC for C++, given the complexities of both the compiler itself and runtime library.  Using the host GCC for C code still makes sense, as `libc`, the C compiler and the C language itself is much more conservative in terms of changes.

These are just my thoughts at the moment and I'm happy to discuss this in more depth (though probably in a different place than this closed PR).

**Pull request closed without merging**